src/v/cloud_storage/materialized_resources.cc

```diff
@@ -379,4 +399,47 @@ void materialized_resources::maybe_trim_segment(
     }
 }
 
+class throttled_dl_source : public ss::data_source_impl {
+public:
+    explicit throttled_dl_source(ss::input_stream<char> st, token_bucket<>& tb)
+      : _src(std::move(st))
+      , _tokens(tb) {}
+
+    ss::future<ss::temporary_buffer<char>> get() override {
+        return ss::with_gate(_gate, [this] {
+            return _src.read().then([this](ss::temporary_buffer<char> buf) {
+                if (!_tokens.try_throttle(buf.size())) {
+                    // TODO: throw different exception
+                    return ss::make_exception_future<
+                      ss::temporary_buffer<char>>(
+                      std::runtime_error("Throttled"));
+                }
+                return ss::make_ready_future<ss::temporary_buffer<char>>(
+                  std::move(buf));
+            });
+        });
+    }
+
+    ss::future<ss::temporary_buffer<char>> skip(uint64_t n) override {
+        return _src.read_exactly(n);
+    }
```

Comment:
Is the intent of this PR to limit network throughput specifically? If so, should we be throttling here, given we're still getting bytes from the data source?